### PR TITLE
add example to test the concept

### PR DESCRIPTION
**Pull request by ClausKlein**
make it works

### CODE REVIEW

File: src/CompilerWarnings.cmake [hunk 1]
```diff
@@ -68,6 +63,13 @@ function(set_project_warnings project_name WARNINGS_AS_ERRORS MSVC_WARNINGS CLAN
     )
   endif()
 
+  message(WARNING "${WARNINGS_AS_ERRORS}")
```
Comment (aminya) [inline src/CompilerWarnings.cmake:66]: A better message for this would be good

File: src/Index.cmake [hunk 2]
```diff
@@ -1,17 +1,19 @@
-cmake_minimum_required(VERSION 3.15)
+cmake_minimum_required(VERSION 3.16)
 
-include(CMakeParseArguments) # to support cmake 3.4 and older
```
Comment (aminya) [inline src/Index.cmake:3]: This comment clarified the intention of this include. In recent versions of Cmake, this is not needed anymore.

File: src/Index.cmake [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
+include(CMakeParseArguments)
 
 set(CMAKELIB_SRC_DIR ${CMAKE_CURRENT_LIST_DIR})
 
+include("${CMAKELIB_SRC_DIR}/PreventInSourceBuilds.cmake")
```
Comment (aminya) [inline src/Index.cmake:7]: Any reason to move this here?

File: .clang-tidy [hunk 4]
```diff
@@ -0,0 +1,41 @@
+---
```
Comment (aminya) [inline .clang-tidy:1]: What's the difference between this and what we used in cpp-best-practices org?

**Review by aminya — CHANGES_REQUESTED** [review]
Looks great. Just a few minor questions.

File: src/CompilerWarnings.cmake (hunk 1 above)
Comment (ClausKlein) [reply src/CompilerWarnings.cmake:66]: I would better remove this message, or not?

File: src/Index.cmake (hunk 2 above)
Comment (ClausKlein) [reply src/Index.cmake:3]: Than the include should be removed; minimum `CMake` version is 3.16

File: src/Index.cmake (hunk 3 above)
Comment (ClausKlein) [reply src/Index.cmake:7]: check early the constrain

File: .clang-tidy (hunk 4 above)
Comment (ClausKlein) [reply .clang-tidy:1]: the extension and the comments, was my fault

**Comment by ClausKlein** [comment]
There are some `cmake-lint` warnings:
```bash
bash-3.2$ grep ' \[W' .cmake-lint.log
CompilerWarnings.cmake:83,25: [W0105] Reference to variable 'project_name' which matches a built-in except for case
Sanitizers.cmake:50,29: [W0105] Reference to variable 'project_name' which matches a built-in except for case
Sanitizers.cmake:51,26: [W0105] Reference to variable 'project_name' which matches a built-in except for case
Tests.cmake:03,27: [W0105] Reference to variable 'project_name' which matches a built-in except for case
Tests.cmake:04,26: [W0105] Reference to variable 'project_name' which matches a built-in except for case
bash-3.2$ 

```

File: .cmake-format.yaml [hunk 5]
```diff
@@ -0,0 +1,135 @@
+parse:
+  additional_commands:
+    cmakelib:
+      pargs:
+        nargs: '*'
+        flags: []
+      spelling: CMakeLib
```
Comment (ClausKlein) [inline .cmake-format.yaml:7]: some convention **warnings** from `cmake-lint`:
```bash
bash-3.2$ grep ' \[C0103' .cmake-lint.log | egrep -vw '(variable|argument)'
Conan.cmake:01,06: [C0103] Invalid macro name "run_conan" doesn't match `[0-9A-Z_]+`
Index.cmake:32,06: [C0103] Invalid macro name "cmakelib" doesn't match `[0-9A-Z_]+`
InterproceduralOptimization.cmake:01,06: [C0103] Invalid macro name "enable_ipo" doesn't match `[0-9A-Z_]+`
PreventInSourceBuilds.cmake:03,09: [C0103] Invalid function name "AssureOutOfSourceBuilds" doesn't match `[0-9a-z_]+`
StaticAnalyzers.cmake:01,06: [C0103] Invalid macro name "enable_cppcheck" doesn't match `[0-9A-Z_]+`
StaticAnalyzers.cmake:18,06: [C0103] Invalid macro name "enable_clang_tidy" doesn't match `[0-9A-Z_]+`
StaticAnalyzers.cmake:30,06: [C0103] Invalid macro name "enable_include_what_you_use" doesn't match `[0-9A-Z_]+`
bash-3.2$ 
```

File: src/Cache.cmake [hunk 6]
```diff
@@ -14,13 +14,17 @@ function(enable_cache)
...
     message(
       STATUS
-        "Using custom compiler cache system: '${CACHE_OPTION}', explicitly supported entries are ${CACHE_OPTION_VALUES}")
+        "Using custom compiler cache system: '${CACHE_OPTION}', explicitly supported entries are ${CACHE_OPTION_VALUES}"
+    )
   endif()
 
-  find_program(CACHE_BINARY ${CACHE_OPTION})
+  #XXX find_program(CACHE_BINARY ${CACHE_OPTION})
+  find_program(CACHE_BINARY NAMES ${CACHE_OPTION_VALUES})
```
Comment (ClausKlein) [inline src/Cache.cmake:22]: IMHO: this is more usable

File: src/Cache.cmake [hunk 7]
```diff
@@ -14,13 +14,17 @@ function(enable_cache)
...
   if(CACHE_BINARY)
-    message(STATUS "${CACHE_OPTION} found and enabled")
-    set(CMAKE_CXX_COMPILER_LAUNCHER ${CACHE_BINARY})
+    message(STATUS "${CACHE_BINARY} found and enabled")
+    set(CMAKE_CXX_COMPILER_LAUNCHER
+        ${CACHE_BINARY}
+        CACHE FILEPATH "compiler cache used")
```
Comment (ClausKlein) [inline src/Cache.cmake:27]: so the user can see and change it ...

**Review by aminya — APPROVED** [review]
Thanks for your contribution

**Comment by aminya** [comment]
I merged this for now as it included some fixes that were necessary. Feel free to send more PRs for more improvements.

**Pull request merged**